File: 0079-word-search/0079-word-search.java
class Solution {

    private static int[][] directions = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

    private boolean dfs(char[][] board, String word, int row, int col, int index){
        if(index==word.length())
            return true;
        if(row < 0 || row >= board.length || col < 0 || col >= board[0].length || word.charAt(index)!=board[row][col])
            return false;
        char temp = board[row][col];
        board[row][col] = '#';
        for(int[] dir: directions){
            int r = row + dir[0];
            int c = col + dir[1];
            if(dfs(board, word, r, c, index + 1))
                return true;
        }
        board[row][col] = temp;
        return false;
    } 

    public boolean exist(char[][] board, String word) {
        for (int row = 0; row<board.length; row++)
            for(int col = 0; col<board[0].length; col++)
                if(dfs(board, word, row, col, 0))
                    return true;
        return false;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROWS, COLS = len(board), len(board[0])
        visited = set()
        def dfs(row, col, i):
            if len(word) == i:
                return True
            if row not in range(ROWS) or col not in range(COLS) or (row, col) in visited or board[row][col] != word[i]:
                return False
            visited.add((row, col))
            res = dfs(row + 1, col, i + 1) or dfs(row - 1, col, i + 1) or dfs(row, col + 1, i + 1) or dfs(row, col - 1, i + 1)
            visited.remove((row, col))
            if res: return True
        for r in range(ROWS):
            for c in range(COLS):
                if dfs(r, c, 0): return True
        return False